### ISSUE TITLE
TemplateNumberName: Incorrect subject number population templates

### ISSUE THREAD

**Issue by MDijsselhof** [ExploreASL/ExploreASL#684]
### Description

In a dataset with subjects that not all include ASL4D.nii's, the amount of scans/subjects is correctly mentioned in the name of the templates present in the Population folder. However, when looking for these templates, xASL still searches for the original nSubjects instead of the correct nSubjects. The data showing the issue is located at: `lood_storage/divi/Projects/ExploreASL/ABBA/NORMENT/StrokeMRI.`

See images below: 

Searching for incorrect name:
![image](https://user-images.githubusercontent.com/75380250/123791190-6e041500-d8df-11eb-8008-515ba0d45c86.png)

Correct name:
![image](https://user-images.githubusercontent.com/75380250/123791264-83793f00-d8df-11eb-8ad7-c630aef3dc03.png)

The origin of the problem is the regular expression (in function `xASL_im_CreateAnalysisMask.m` line 144): 

` PathFoV = xASL_adm_GetFileList(x.D.TemplatesStudyDir, ['^FoV_n' xASL_num2str(x.dataset.nSubjectsSessions) '_bs-mean\.nii$'], 'FPList', [1 1]);` 

The name of the file is defined in `xASL_im_CreateAnalysisMask.m`, where in the subfunction `xASL_stat_RobustMean` empty, corrupt and outlier images are removed in the cell-structure. The length of the remaining cell structure is then used to determine the remaining files from which a template is made and the file is assigned that name in: ` NameIM = [TemplateNameList{iScanType} '_n' num2str(length(NotOutliers))];` 

Proposed solution:
A) in this issue, I fix that the name created in the templates is stored n the x-struct and reused throughout the pipeline. 
B) we create a general provenance stamp in the x-struct/json: the ExploreASL version and date-time stamp. When ExploreASL is rerun with a different version than the x-struct/json has, it will issue a very clear warning. 

Future issues:
We would need to check if the x-struct is always saved and reloaded for provenance, not to rely only on memory (for pipeline crashing). Note that the population module doesn't have to run in parallel.
B) 

### Requirements

- [x] replace `x.datasets.nSubjectsSessions` by `x.S.TemplateNumberName` that can be used throughout the population module
Note that the number of sessions is still not nicely processed, so we keep the number of subjects only throughout
- [x] replace `x.nSubjects` by `x.dataset.nSubjects` (to sync with where also other n parameter-fields live, like `x.dataset.nSessions` etc)
- [x] Make sure that already the first lines of xASL_module_Population have x.S.Atlases field. `xASL_module_Population` like 48 initializes Atlases field. But at this point, it should be already read from the dataPar.json.

### How to test

@MDijsselhof can you please test this? I couldn't test this with your folder, which only contains 1 subject (templates are only created with more than 6 subjects, as you need a minimal sample size to create a robust template)

### Release notes

Fix issue with subject number in population filenames.

**Comment by HenkMutsaerts** [comment]
This was programmed because it could take a long time to recreate templates.

@MDijsselhof would it be a solution to replace the n-name by a date and time string that is created at the start of the population module, which is then remembered throughout the population statistics?

This is then always created and this won't crash. I already programmed something similar for the import overview file; which I can take as example.

**Comment by MDijsselhof** [comment]
Yes this will resolve the issue as far as I can tell

**Comment by HenkMutsaerts** [comment]
Let's generalize this.
1) We want that the name of the current processing travels through the whole population module; such that the correct files (e.g., templates) are used in the remainder of the population module. This is a MUST-HAVE.
2) It is a good idea to add a date-time stamp to the file name to know when a template was created (NICE2HAVE)
3) It takes a long-time to recreate the templates, so it would be useful to re-use templates that were created previously; but we would like to know that they were with a similar analysis, i.e., similar sample size (MUSTHAVE) and similar analysis (ExploreASL version, NICE2HAVE)

Proposed solution:
A) in this issue, I fix that the name created in the templates is stored n the x-struct and reused throughout the pipeline. 
B) we create a general provenance stamp in the x-struct/json: the ExploreASL version and date-time stamp. When ExploreASL is rerun with a different version than the x-struct/json has, it will issue a very clear warning. 


Future issues:
We would need to check if the x-struct is always saved and reloaded for provenance, not to rely only on memory (for pipeline crashing). Note that the population module doesn't have to run in parallel.
B)

**Comment by MDijsselhof** [comment]
![image](https://user-images.githubusercontent.com/75380250/213180486-c876ed25-1350-428d-8424-6e3f536abf5f.png)

I get the following error when reprocessing the StrokeMRI dataset (after removing only the population module lock files).

**Comment by jan-petr** [comment]
I have tried to run ExploreASL(path, 0, [0 0 1]). And already at line 38 in `xASL_module_Population`, I have x.S.Atlases that must have been obtained from dataPar.json - I have tried modifying the dataPar.json and it was already correctly read at this point of code.

I have moved the check for atlases higher in the code. But that excluded, it should have worked with the `dataPar` right from the start.
